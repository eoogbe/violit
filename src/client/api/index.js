/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { load, loadAll } from './load';
export { default as InvalidResponseError } from './InvalidResponseError';
export { default as checkExists } from './checkExists';
export { default as checkPermissions } from './checkPermissions';
export { default as href } from './href';
export { default as sendCreate } from './sendCreate';
export { default as sendDelete } from './sendDelete';
export { default as sendPost } from './sendPost';
