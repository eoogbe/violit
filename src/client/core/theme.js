/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export default {
  primaryColor: '#7F00FF',

  gray200: '#EEEEEE',
  gray300: '#E0E0E0',
  gray500: '#9E9E9E',
  gray700: '#616161',

  red200: '#EF9A9A',
  red500: '#F44336',

  lightBlue700: '#0288D1',

  black: '#000000',
  white: '#FFFFFF',

  spacingSm: '0.25em',
  spacingMd: '0.5em',
  spacingLg: '1em',

  borderRadius: '2px',
};
