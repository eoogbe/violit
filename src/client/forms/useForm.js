/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import camelCase from 'camel-case';
import * as R from 'ramda';
import { useCallback, useEffect, useMemo, useRef, useState } from 'react';

/** Hook that adds state for a form with controlled components. */
export default (initialValues, validate, options) => {
  const { commitOnChange } = options || {};

  const prevValues = useRef(initialValues);
  const preventDoubleSubmit = useRef(false);
  const [values, setValues] = useState(initialValues);
  const [errors, setErrors] = useState({});
  const [dirtied, setDirtied] = useState(R.map(R.F, initialValues));
  const [submitting, setSubmitting] = useState(false);

  const unmounted = useRef(false);
  useEffect(() => () => {
    unmounted.current = true;
  }, []);

  const isDirty = useMemo(() => {
    return Object.values(dirtied).includes(true);
  }, [dirtied]);

  const hasError = useMemo(() => {
    return Object.values(errors).some((error) => !!error);
  }, [errors]);

  const validateFields = useCallback(async () => {
    if (!validate) {
      return {};
    }

    const errors = await validate(values) || {};
    if (!unmounted.current) {
      setErrors(errors);
    }

    return errors;
  }, [validate, values]);

  useEffect(() => {
    if (values === prevValues.current) {
      return;
    }

    validateFields();
    prevValues.current = values;
  }, [values, validateFields]);

  const reset = useCallback(() => {
    if (unmounted.current) {
      return;
    }

    prevValues.current = initialValues;
    preventDoubleSubmit.current = false;
    setValues(initialValues);
    setErrors({});
    setDirtied(R.map(R.F, initialValues));
    setSubmitting(false);
  }, [initialValues]);

  const handleInput = useCallback(({ target }) => {
    const { name, value } = target;
    setValues(R.assoc(camelCase(name), value));
  }, []);

  const handleCommit = useCallback(({ target }) => {
    const { value } = target;
    const name = camelCase(target.name);
    setValues(R.assoc(name, value));
    setDirtied(R.assoc(name, true));
  }, []);

  const handleChange = useCallback((event) => {
    handleInput(event);
    if (commitOnChange) {
      handleCommit(event);
    }
  }, [handleInput, handleCommit, commitOnChange]);

  const handleSubmit = useCallback((onSubmit) => async (event) => {
    event.preventDefault();

    if (preventDoubleSubmit.current) {
      return;
    }

    if (!unmounted.current) {
      preventDoubleSubmit.current = true;
      setSubmitting(true);
    }

    const fieldErrors = await validateFields();
    if (Object.values(fieldErrors).every((error) => !error)) {
      const submitErrors = await onSubmit(values) || {};
      if (!unmounted.current) {
        setErrors(submitErrors);
      }
    }

    if (!unmounted.current) {
      preventDoubleSubmit.current = false;
      setSubmitting(false);
    }
  }, [validateFields, values]);

  const eventHandlers = {
    handleChange,
    handleBlur: handleCommit,
    handleSubmit,
  };

  return [
    values,
    errors,
    dirtied,
    eventHandlers,
    { submitting, isDirty, hasError, reset, validateFields },
  ];
};
