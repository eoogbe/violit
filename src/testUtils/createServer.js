/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import cookieParser from 'cookie-parser';
import express from 'express';
import routes from 'server/api/routes';

/** Creates a test server for the specified `router`. */
export default (router, locals) => {
  const app = express();
  app.use(cookieParser());
  app.use(express.json());
  app.use((req, res, next) => {
    Object.assign(res.locals, { routes }, locals);
    next();
  });
  app.use(router());
  app.use((err, req, res, next) => {
    res.status(500).json({ error: err.message });
  });
  return app;
};
