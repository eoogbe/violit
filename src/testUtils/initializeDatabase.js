/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import path from 'path';
import Database from 'better-sqlite3';
import bcrypt from 'bcrypt';
import * as BOARD from '__fixtures__/boardFixture';
import * as COMMENTS from '__fixtures__/commentsFixture';
import * as THREADS from '__fixtures__/threadsFixture';
import * as USER from '__fixtures__/userFixture';

const INSERT_USER_SQL =
`INSERT INTO user (id, username, password)
VALUES (:id, :username, :password)`;

const INSERT_BOARD_SQL =
`INSERT INTO board (id, name)
VALUES (:id, :name)`;

const INSERT_THREAD_SQL =
`INSERT INTO thread (
  id,
  slug,
  dateCreated,
  dateDeleted,
  headline,
  articleBody,
  userId,
  boardId
) VALUES (
  :id,
  :slug,
  :dateCreated,
  :dateDeleted,
  :headline,
  :articleBody,
  :userId,
  :boardId
)`;

const INSERT_COMMENT_SQL =
`INSERT INTO comment (id, dateCreated, dateDeleted, content, userId, threadId)
VALUES (:id, :dateCreated, :dateDeleted, :content, :userId, :threadId)`;

/** Initializes the test database. */
export default () => {
  const dbPath = path.resolve(__dirname, '..', '..', 'data', 'test.db');
  const db = new Database(dbPath, {
    fileMustExist: true,
  });

  db.prepare('DELETE FROM comment').run();
  db.prepare('DELETE FROM thread').run();
  db.prepare('DELETE FROM board').run();
  db.prepare('DELETE FROM user').run();

  db.prepare(INSERT_USER_SQL).run({
    id: 'user1',
    username: USER.username,
    password: bcrypt.hashSync(USER.password, 4),
  });

  db.prepare(INSERT_BOARD_SQL).run(BOARD.board);

  const insertThreadStmt = db.prepare(INSERT_THREAD_SQL);
  THREADS.db.forEach((thread) => {
    insertThreadStmt.run(thread);
  });

  const insertCommentStmt = db.prepare(INSERT_COMMENT_SQL);
  COMMENTS.db.forEach((comment) => {
    insertCommentStmt.run(comment);
  });

  return db;
};
