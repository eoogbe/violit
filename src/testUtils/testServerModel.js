/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import * as BOARD from '__fixtures__/boardFixture';
import * as COMMENTS from '__fixtures__/commentsFixture';
import * as THREADS from '__fixtures__/threadsFixture';
import * as USER from '__fixtures__/userFixture';

const CAN_DELETE_THREAD_RESULTS = {
  [THREADS.thread1.slug]: true,
  [THREADS.thread2.slug]: false,
  [THREADS.thread3.slug]: true,
};

const CAN_DELETE_COMMENT_RESULTS = {
  comment1: true,
  comment2: false,
  comment3: true,
};

/** Creates a test server model with default values. */
export const createModel = R.mergeDeepRight({
  auth: {
    sign: () => Promise.resolve('token123'),
    getCredentials: () => Promise.resolve(null),
    createCredentials: () => Promise.resolve('token123'),
  },
  board: {
    getBoard: () => BOARD.board,
  },
  threads: {
    getThreads: () => THREADS.unnormalized,
    getThreadById: (id) => ({
      id,
      slug: THREADS.thread1.slug,
      dateCreated: THREADS.thread1.dateCreated,
      headline: THREADS.thread1.headline,
      articleBody: THREADS.thread1.articleBody,
      author: THREADS.thread1.author,
    }),
    getThreadBySlug: (slug) => ({
      slug,
      id: 'thread1',
      dateCreated: THREADS.thread1.dateCreated,
      headline: THREADS.thread1.headline,
      articleBody: THREADS.thread1.articleBody,
      author: THREADS.thread1.author,
    }),
    createThread: () => THREADS.normalized.thread1,
    softDeleteThread: () => ({
      id: 'thread1',
      slug: THREADS.thread1.slug,
      dateCreated: THREADS.thread1.dateCreated,
      dateDeleted: THREADS.thread1.dateDeleted,
      headline: THREADS.thread1.headline,
      articleBody: THREADS.thread1.articleBody,
      author: THREADS.thread1.author,
    }),
    canDeleteThread: (thread) => CAN_DELETE_THREAD_RESULTS[thread.slug],
  },
  comments: {
    getCommentsByThreadSlug: () => COMMENTS.unnormalized,
    getCommentById: (id) => ({
      ...COMMENTS.normalized.comment1,
      id,
    }),
    createComment: () => COMMENTS.normalized.comment1,
    softDeleteComment: (id) => ({
      id,
      dateCreated: COMMENTS.comment1.dateCreated,
      dateDeleted: COMMENTS.comment1.dateDeleted,
      text: COMMENTS.comment1.text,
      threadSlug: COMMENTS.comment1.threadSlug,
      author: COMMENTS.comment1.author,
    }),
    canDeleteComment: (comment) => CAN_DELETE_COMMENT_RESULTS[comment.id],
  },
  users: {
    getUserById: () => USER.credentials,
    getUserByUsername: (username) => ({
      ...USER.credentials,
      username,
    }),
    createUser: () => Promise.resolve(USER.credentials),
  },
});

/** Creates a test server model for an authenticated user. */
export const createAuthenticatedModel =
  R.mergeDeepRight(createModel({
    auth: {
      getCredentials: () => Promise.resolve(USER.credentials),
    },
  }));
