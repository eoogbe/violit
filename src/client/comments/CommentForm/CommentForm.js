/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Button } from 'client/core';
import { Input, useForm } from 'client/forms';

const StyledCommentForm = styled.form`
  font-size: 1rem;
  display: flex;
  flex-direction: column;
  margin: ${({ theme }) => theme.spacingLg} 0 ${({ theme }) => theme.spacingMd};
`;

const FieldList = styled.ul`
  margin: 0;
  padding: 0;
  list-style: none;
`;

const SubmitButton = styled(Button)`
  align-self: flex-end;
`;

/** A form for creating a comment. */
const CommentForm = ({ createComment }) => {
  const [
    values,
    /* errors */,
    /** dirtied */,
    { handleChange, handleSubmit },
    { submitting, isDirty, hasError, reset },
  ] = useForm({
    text: '',
  }, ({ text }) => ({
    text: !text.trimEnd() && 'Must not be blank.',
  }), {
    commitOnChange: true,
  });
  const disabled = !isDirty || hasError || submitting;

  const onSubmit = (input) => {
    createComment(input);
    reset();
  };

  return (
    <StyledCommentForm onSubmit={handleSubmit(onSubmit)}>
      <FieldList>
        <li>
          <Input
            as="textarea"
            name="text"
            placeholder="Write a comment"
            rows={5}
            value={values.text}
            onChange={handleChange}
          />
        </li>
      </FieldList>
      <SubmitButton type="submit" disabled={disabled} loading={submitting}>
        Comment
      </SubmitButton>
    </StyledCommentForm>
  );
};

CommentForm.propTypes = {
  /** Creates a comment. */
  createComment: PropTypes.func.isRequired,
};

export default CommentForm;
