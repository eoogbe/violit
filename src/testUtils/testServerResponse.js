/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import routes from 'server/api/routes';
import { createAuthenticatedModel, createModel } from './testServerModel';
import * as USER from '__fixtures__/userFixture';

/** Creates a test response object with default values. */
export const createResponse = R.mergeDeepRight({
  status: jest.fn().mockReturnThis(),
  set: jest.fn().mockReturnThis(),
  clearCookie: jest.fn().mockReturnThis(),
  send: jest.fn(),
  locals: {
    model: createModel({}),
    routes,
  },
});

/** Creates a test response object for an authenticated user. */
export const createAuthenticatedResponse = R.mergeDeepRight(createResponse({
  locals: {
    model: createAuthenticatedModel({}),
    credentials: USER.credentials,
  },
}));
