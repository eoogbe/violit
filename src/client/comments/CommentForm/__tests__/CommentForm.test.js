/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import TestRenderer from 'react-test-renderer';
import { Button } from 'client/core';
import { Input } from 'client/forms';
import { updateWrapper } from 'testUtils/wait';
import CommentForm from '../CommentForm';

const changeText = (wrapper, value) => {
  wrapper.find(Input).simulate('change', {
    target: { name: 'text', value },
  });
};

const simulateSubmit = (wrapper) => {
  const onSubmit = wrapper.find(CommentForm).children().prop('onSubmit');
  return act(async () => {
    await onSubmit({
      preventDefault: () => {},
    });
    await updateWrapper(wrapper);
  });
};

it('renders without crashing', () => {
  const renderer = TestRenderer.create(<CommentForm createComment={() => {}} />);
  expect(renderer.toJSON()).toMatchSnapshot();
});

it('when submitting triggers comment creation', async () => {
  const createComment = jest.fn();

  const wrapper = mount(<CommentForm createComment={createComment} />);
  expect(wrapper.find(Button).prop('disabled')).toBe(true);

  changeText(wrapper, 'Lorem ipsum  ');

  await simulateSubmit(wrapper);

  expect(createComment).toHaveBeenCalled();
  expect(wrapper.find(Input).prop('value')).toBe('');
  expect(wrapper.find(Button).prop('disabled')).toBe(true);
});

it('with blank comment disables the submit button', async () => {
  const wrapper = mount(<CommentForm createComment={() => {}} />);

  changeText(wrapper, '  ');

  await updateWrapper(wrapper);

  expect(wrapper.find(Button).prop('disabled')).toBe(true);
});
