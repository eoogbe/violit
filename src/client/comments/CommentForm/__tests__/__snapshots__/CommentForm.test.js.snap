// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<form
  className="sc-dnqmqq jjuQyM"
  onSubmit={[Function]}
>
  <ul
    className="sc-iwsKbI dgXfCH"
  >
    <li>
      <textarea
        className="sc-htoDjs kEmxlR"
        name="text"
        onChange={[Function]}
        placeholder="Write a comment"
        rows={5}
        value=""
      />
    </li>
  </ul>
  <button
    className="sc-ifAKCX hbBAat sc-gZMcBi exoESG"
    disabled={true}
    type="submit"
  >
    Comment
  </button>
</form>
`;
